DylibBinder/DBGenericParameter.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections.Generic;
+using Dynamo.SwiftLang;
+using SwiftReflector;
+using SwiftReflector.Inventory;
+
+namespace DylibBinder {
+	internal class DBGenericParameter : IEquatable<DBGenericParameter> {
+		public DBGenericParameter (int depth, int index)
+		{
+			Depth = depth;
+			Index = index;
+			Name = SLGenericReferenceType.DefaultNamer (depth, index);
+		}
+
+		public int Depth { get; }
+		public int Index { get; }
+		public string Name { get; }
+
+		public bool Equals (DBGenericParameter other)
+		{
+			return this.Depth == other.Depth && this.Index == other.Index;
+		}
+	}
+
+	internal class DBGenericParameters {
+
+		public DBGenericParameters (SwiftBaseFunctionType signature) {
+			AssignGenericParameters (signature);
+		}
+
+		public DBGenericParameters (SwiftType swiftType)
+		{
+			FilterGenericParameterSource (swiftType);
+		}
+
+		public DBGenericParameters (DBTypeDeclaration typeDeclaration)
+		{
+			ParseTopLevelGenerics (typeDeclaration.Funcs, typeDeclaration.Properties);
+		}
+
+		public List<DBGenericParameter> GenericParameters { get; } = new List<DBGenericParameter> ();
+
+
+		void AssignGenericParameters (SwiftBaseFunctionType signature)
+		{
+			FilterGenericParameterSource (signature.Parameters, signature.ReturnType);
+		}
+
+		void FilterGenericParameterSource (params SwiftType [] types)
+		{
+			foreach (var type in types) {
+				switch (type) {
+				case SwiftFunctionType funcType:
+					HandleType (funcType);
+					return;
+				case SwiftBoundGenericType boundType:
+					HandleType (boundType);
+					return;
+				case SwiftTupleType tupleType:
+					HandleType (tupleType);
+					return;
+				case SwiftGenericArgReferenceType refType:
+					HandleType (refType);
+					return;
+				default:
+					return;
+				}
+			}
+		}
+
+		void HandleType (SwiftFunctionType funcType)
+		{
+			switch (funcType.Parameters) {
+			case SwiftBoundGenericType boundType:
+				HandleType (boundType);
+				return;
+			case SwiftTupleType tupleType:
+				HandleType (tupleType);
+				return;
+			case SwiftGenericArgReferenceType refType:
+				HandleType (refType);
+				return;
+			default:
+				return;
+			}
+		}
+
+		void HandleType (SwiftBoundGenericType boundType)
+		{
+			foreach (var bound in boundType.BoundTypes) {
+				if (bound is SwiftGenericArgReferenceType refType)
+					HandleType (refType);
+			}
+		}
```

Comment:
All this switches can be written in a much nicer way if we change the API and rather than having a method that adds to the list, we return the DBGenericParameter and add it, something like:
```csharp
    DBGenericParameter GetDBParameter (SwiftBoundGenericType t) {
        return null;
    }
    
    DBGenericParameter GetDBParameter (SwiftTupleType t) {
        return null;
    }
    
    DBGenericParameter GetDBParameter (SwiftGenericArgReferenceType t) {
        return null;
    }


    DBGenericParameter GetDBParameter (SwiftType funcType) => funcType switch {
            SwiftBoundGenericType boundType => GetDBParameter (boundType),
	    SwiftTupleType tupleType => GetDBParameter (tupleType),
            SwiftGenericArgReferenceType refType => GetDBParameter (refType),
            _ => throw new ArgumentOutOfRangeException()
    };
```
We should rename Handle to something nicer, like GetDBParameter or something and then do:
```csharp
collection.Add (GetDBParameter (functType) );
```